language/move-prover/diagen/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use log::{error, info};
+use regex::Regex;
+use serde::{self, Deserialize, Serialize};
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet, VecDeque},
+    env, fs,
+    io::{self, Write},
+    path::{Path, PathBuf},
+    process::{Command, Stdio},
+};
+
+/// Options passed into the documentation generator.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct DiagenOptions {
+    pub generate_dot: bool,
+    pub generate_pdf: bool,
+    pub generate_svg: bool,
+    pub generate_global_graph: bool,
+}
+
+impl Default for DiagenOptions {
+    fn default() -> Self {
+        Self {
+            generate_dot: false,
+            generate_pdf: false,
+            generate_svg: false,
+            generate_global_graph: false,
+        }
+    }
+}
+
+pub fn move_files_in(path: &PathBuf) -> Vec<PathBuf> {
+    fs::read_dir(path)
+        .unwrap()
+        .filter_map(|dir_entry| {
+            let path = dir_entry.unwrap().path();
+            if path.is_file() && path.extension().is_some() && path.extension().unwrap() == "move" {
+                Some(path)
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+// Locate the path of the "language" directory in the Libra repository.
+pub fn locate_language_dir() -> io::Result<PathBuf> {
```

Comment:
Docgen is a general tool which may be run outside of the libra tree. Even if its in the Libra tree, the location of diagen may change. This here is not very clean. It crashes if somewhere else in the file system. As said, the path to where to store diagen should be passed into a function in this crate, then this might not be longer necessary.

Reply:
Refactored accordingly.